pydantic/main.py

```diff
@@ -238,7 +249,46 @@ def __new__(mcs, name, bases, namespace, **kwargs):  # noqa C901
             '__custom_root_type__': _custom_root_type,
             **{n: v for n, v in namespace.items() if n not in fields},
         }
-        return super().__new__(mcs, name, bases, new_namespace, **kwargs)
+
+        cls = super().__new__(mcs, name, bases, new_namespace, **kwargs)
+
+        # Generating right signature for __init__:
+        orig_init = cls.__init__
+        orig_code = orig_init.__code__
+        # cannot modify cls.__init__ directly,
+        # as it shared between children classes, so we copy it
+        cls.__init__ = new_init = FunctionType(
+            orig_code, orig_init.__globals__, closure=orig_init.__closure__, name=orig_init.__name__
+        )
+
+        orig_posonlycount = orig_code.co_posonlyargcount if PY38 else 0
+        orig_argnames = orig_code.co_varnames[: orig_code.co_argcount + orig_code.co_kwonlyargcount + orig_posonlycount]
+        orig_argnames_set = set(orig_argnames)
+        fields_names = tuple([name for name in fields if name not in orig_argnames_set])
+        del orig_argnames_set
+
+        def fake_init():
+            ...
+
+        fake_init = FunctionType(
+            copy_code(
+                fake_init.__code__,
+                co_argcount=orig_code.co_argcount,
+                co_kwonlyargcount=orig_code.co_kwonlyargcount + len(fields_names),
+                co_names=orig_code.co_names + fields_names,
+                co_varnames=orig_argnames + fields_names,
+                co_name=orig_code.co_name,
+                **({'co_posonlyargcount': orig_posonlycount} if PY38 else {}),
+            ),
+            orig_init.__globals__,
+            name=orig_init.__name__,
+        )
+        fake_init.__doc__ = new_init.__doc__ = orig_init.__doc__
```

Comment:
Would it be a good idea to add a helpful docstring to the real `__init__` which by default appears in `help(MyModel.__init__)`?

Reply:
It’s not bad idea for sure. What should be written there?

Reply:
Not sure, something like:

> Create a new model by parsing and validating input data from keyword arguments, a `ValidationError` will be raised if the input data cannot be parsed to for a valid model.

Or something.

Reply:
Also adding info about original signature to docstring on generation